executor/grant.go

```diff
@@ -246,30 +248,31 @@ func (e *GrantExec) grantDBPriv(priv *ast.PrivElem, user *ast.UserSpec) error {
 		return errors.Trace(err)
 	}
 	userName, host := parseUser(user.User)
-	sql := fmt.Sprintf(`UPDATE %s.%s SET %s WHERE User="%s" AND Host="%s" AND DB="%s";`, mysql.SystemDB, mysql.DBTable, asgns, userName, host, db.Name.O)
+	sql := fmt.Sprintf(`UPDATE %s.%s SET %s WHERE User="%s" AND Host="%s" AND DB="%s";`, mysql.SystemDB, mysql.DBTable, asgns, userName, host, dbName)
 	_, err = e.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(e.ctx, sql)
 	return errors.Trace(err)
 }
 
 // Manipulate mysql.tables_priv table.
 func (e *GrantExec) grantTablePriv(priv *ast.PrivElem, user *ast.UserSpec) error {
-	db, tbl, err := e.getTargetSchemaAndTable()
+	dbName, err := e.getTargetSchema()
 	if err != nil {
 		return errors.Trace(err)
 	}
+	tblName := e.Level.TableName
 	userName, host := parseUser(user.User)
-	asgns, err := composeTablePrivUpdate(e.ctx, priv.Priv, userName, host, db.Name.O, tbl.Meta().Name.O)
+	asgns, err := composeTablePrivUpdate(e.ctx, priv.Priv, userName, host, dbName, tblName)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	sql := fmt.Sprintf(`UPDATE %s.%s SET %s WHERE User="%s" AND Host="%s" AND DB="%s" AND Table_name="%s";`, mysql.SystemDB, mysql.TablePrivTable, asgns, userName, host, db.Name.O, tbl.Meta().Name.O)
+	sql := fmt.Sprintf(`UPDATE %s.%s SET %s WHERE User="%s" AND Host="%s" AND DB="%s" AND Table_name="%s";`, mysql.SystemDB, mysql.TablePrivTable, asgns, userName, host, dbName, tblName)
```

Comment:
When I want to grant privilege of all database that begin with name 'te%' to someone, 
it doesn't insert multiple records, but just one, into mysql.db, like this:

```
+------+-----+--------+-------------+-------------+-------------+-------------+-------------+-----------+------------+------------+------------+--------------+
| Host | DB  | User   | Select_priv | Insert_priv | Update_priv | Delete_priv | Create_priv | Drop_priv | Grant_priv | Index_priv | Alter_priv | Execute_priv |
+------+-----+--------+-------------+-------------+-------------+-------------+-------------+-----------+------------+------------+------------+--------------+
| %    | te% | genius | Y           | Y           | Y           | Y           | Y           | Y         | Y          | Y          | Y          | Y            |
+------+-----+--------+-------------+-------------+-------------+-------------+-------------+-----------+------------+------------+------------+--------------+
``` 

In the grant logic, this kind of record will be checked, if not exist, create one first.
Update will update this record, and use exactly `=`.

When used for privilege check, this record will be Select use `like` pattern match, 
so we know all database that begin with name 'te%' apply the rule of privilege in the record.